src/observer.ts

```diff
@@ -361,22 +367,49 @@ export class DataLayerObserver {
     }
 
     try {
+      // NOTE a target that is not found will result in an Error being thrown
       const target = DataLayerTarget.find(source);
-      this.registerTarget(target, operators, destination, readOnLoad, monitor, debug, debounce);
-    } catch (_) {
-      // schedule subsequent attempts at (attempt * wait) later
-      setTimeout(() => {
-        try {
-          const target = DataLayerTarget.find(source);
-          this.registerTarget(target, operators, destination, readOnLoad, monitor, debug, debounce);
-        } catch (err) {
-          if (attempt > 3) {
-            Logger.getInstance().error(LogMessageType.RuleRegistrationError, { rule: id, source, reason: err.message });
-          } else {
-            this.registerRule(rule, attempt + 1);
-          }
+
+      // if `waitUntil`, a value in milliseconds or predicate test is expected to delay registration
+      if (waitUntil) {
+        switch (typeof waitUntil) {
+          case 'number':
+            // NOTE this delay is set *after* the data layer is found to be defined on the page
+            setTimeout(() => {
+              // re-register the rule so that it schedules immediately upon next invocation of `registerRule`
+              this.registerRule({
+                ...rule,
+                waitUntil: 0,
+              }, attempt);
+            }, waitUntil > -1 ? waitUntil : 0); // error check a negative value and schedule immediately if used
+            break;
+          case 'function':
+            if (!waitUntil(target.value)) {
+              // NOTE this is handled and rescheduled by the `catch` block
+              throw Error('Rule predicate test is falsy');
```

Comment:
wouldn't `waitUntil` return a false value if the predicate invocation returned false? I.e. if the test is the existence of an object, and that object doesn't exist yet (b/c it hasn't loaded) then you'd expect `waitUntil` to return false - but this  logic treats a false value as an exception. I think I'm getting tripped up b/c this logic uses error handling to handle expected use-cases. It might be cleaner/more interpretable if the `setTimeout` functions were handled in the main `try` block. WDYT?

Reply:
I agree.  Let me revisit with this context.